layers: Added VK_NV_framebuffer_mixed_samples

cmayer1 opened the pull request:
Added support for VK_NV_framebuffer_mixed_samples to the validation
layers.

Test the following VUIDs in case that the extension is present:
VUID-VkGraphicsPipelineCreateInfo-subpass-01411
VUID-VkGraphicsPipelineCreateInfo-subpass-01412
VUID-VkPipelineMultisampleStateCreateInfo-rasterizationSamples-01415
VUID-VkPipelineCoverageModulationStateCreateInfoNV-
coverageModulationTableEnable-01405

Test the following VUIDs in case that the extension is not present:
VUID-VkGraphicsPipelineCreateInfo-subpass-00757

Added tests FramebufferMixedSamplesNV and FramebufferMixedSamples
to verify that the relevant VUIDs are touched.

Changed CmdClearAttachmentTests to use VK_SAMPLE_COUNT_1_BIT for
rasterization samples to avoid running into
VUID-VkGraphicsPipelineCreateInfo-subpass-00757.

quadgod420 commented:
I can verify this commit fixes the issue I reported when using VK_NV_framebuffer_mixed_samples https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/390. Thanks for adding this :)

cmayer1 commented:
@jzulauf-lunarg 
Hi John,
I created this pull request some days ago but so far it was not yet assigned to a reviewer. Sorry for contacting you directly but since this is my first submission to this repository I want to make sure I did not miss anything to make the pull request visible to reviewers. 

Thanks for any feedback

jzulauf-lunarg commented:
@cmayer1  -- my apologies.  Please feel free to ping me or tag me in future submissions.

jzulauf-lunarg commented on layers/core_validation.cpp:
```diff
@@ -1638,6 +1640,31 @@ static bool ValidatePipelineUnlocked(layer_data *dev_data, std::vector<std::uniq
+            subpass_num_samples |= static_cast<uint32_t>(pPipeline->rp_state->createInfo.pAttachments[attachment].samples);
+        }
+
+        if (IsPowerOfTwo(subpass_num_samples) && subpass_num_samples != raster_samples) {
```
Wouldn't this condition be `!IsPowerOfTwo(subpass_num_samples) || (subpass_num_samples != raster_samples)` ?

Not  being power of two means that more than one sample rate is being used, (and thus can't match a valid single bit rasterizationSamples value.

cmayer1 replied:
Changed it and added a check to make sure num_samples is greater than 0 to cover the case of not having any color or depth attachment.

jzulauf-lunarg commented on layers/core_validation.cpp:
```diff
@@ -1664,6 +1691,73 @@ static bool ValidatePipelineUnlocked(layer_data *dev_data, std::vector<std::uniq
+        uint32_t raster_samples = static_cast<uint32_t>(GetNumSamples(pPipeline));
+        uint32_t subpass_color_samples = 0;
+
+        for (uint32_t i = 0; i < subpass_desc->colorAttachmentCount; i++) {
```
Is this the exact same loop as above?  If so, factor into function?

cmayer1 replied:
yes, they are the same. I added a lambda function.

jzulauf-lunarg commented on layers/core_validation.cpp:
```diff
@@ -1664,6 +1691,73 @@ static bool ValidatePipelineUnlocked(layer_data *dev_data, std::vector<std::uniq
+                                             (pPipeline->graphicsPipelineCI.pDepthStencilState->depthBoundsTestEnable == VK_TRUE) ||
+                                             (pPipeline->graphicsPipelineCI.pDepthStencilState->stencilTestEnable == VK_TRUE);
+
+                if (ds_test_enabled && (raster_samples != subpass_depth_samples)) {
```
Here we don't do a pow2 check (which for *valid* raster_samples* is irrelevant, as valid rasters_samples must be pow2) -- a consistent check would be more maintainable.

cmayer1 replied:
Done

jzulauf-lunarg commented on layers/core_validation.cpp:
```diff
@@ -1664,6 +1691,73 @@ static bool ValidatePipelineUnlocked(layer_data *dev_data, std::vector<std::uniq
+            }
+        }
+
+        if (IsPowerOfTwo(subpass_color_samples)) {
```
Not sure why we aren't testing for any of the VU's when subpass_color_samples aren't pow2 (i.e. have more than one sample rate across attachments).

jzulauf-lunarg replied:
looks like we should be comparing with the MSB of subpass_color_samples (or gathering a subpass_max_color_samples in that loop)

cmayer1 replied:
My understanding is that the sample count of  the different color targets should be identical. That is why I only test the VUs if subpass_color_samples is a power of two.  If my assumption is wrong, I will change it.

jzulauf-lunarg commented on layers/core_validation.cpp:
```diff
@@ -1664,6 +1691,73 @@ static bool ValidatePipelineUnlocked(layer_data *dev_data, std::vector<std::uniq
+                        HandleToUint64(pPipeline->pipeline), "VUID-VkPipelineMultisampleStateCreateInfo-rasterizationSamples-01415",
+                        "vkCreateGraphicsPipelines: pCreateInfo[%d].pMultisampleState->sampleShadingEnable must be VK_FALSE.",
+                        pipelineIndex);
+                }
```
Error message seems incomplete.  

"must be VK_FALSE, when ...CreateInfo.rasterizationSamples > subpass_color_samples" 

Perhaps?

cmayer1 replied:
Done

jzulauf-lunarg commented on tests/layer_validation_tests.cpp:
```diff
@@ -33966,6 +33966,201 @@ TEST_F(VkLayerTest, InlineUniformBlockEXT) {
+        rpi.pSubpasses = &sp;
+
+        VkRenderPass rp;
+        m_errorMonitor->SetDesiredFailureMsg(VK_DEBUG_REPORT_ERROR_BIT_EXT,
```
Usually for VerifyNotFound cases either m_errorMonitor->ExpectSuccess(); or no call to  SetDesired .. so guard with the same condition as below.

cmayer1 replied:
Done

jzulauf-lunarg commented on tests/layer_validation_tests.cpp:
```diff
@@ -16737,7 +16737,7 @@ TEST_F(VkLayerTest, CmdClearAttachmentTests) {
     pipe_ms_state_ci.sType = VK_STRUCTURE_TYPE_PIPELINE_MULTISAMPLE_STATE_CREATE_INFO;
     pipe_ms_state_ci.pNext = NULL;
-    pipe_ms_state_ci.rasterizationSamples = VK_SAMPLE_COUNT_4_BIT;
+    pipe_ms_state_ci.rasterizationSamples = VK_SAMPLE_COUNT_1_BIT;
```
unless it make the code non-bisectable, test changes should be in a subsequent commit to layer functionality changes

cmayer1 replied:
Done

cmayer1 commented on layers/core_validation.cpp:
```diff
@@ -1327,6 +1327,8 @@ static bool ValidatePipelineLocked(layer_data *dev_data, std::vector<std::unique
     return skip;
 }
 
+static bool IsPowerOfTwo(unsigned x) { return x && !(x & (x - 1)); }
```
Done

cmayer1 commented:
Thanks a lot for reviewing!
I addressed most of your suggestions and created a new pull request: https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/481
